packages/task/src/browser/task-configuration-manager.ts

```diff
@@ -0,0 +1,200 @@
+
+import debounce = require('p-debounce');
+import { inject, injectable, postConstruct } from 'inversify';
+import URI from '@theia/core/lib/common/uri';
+import { Emitter, Event } from '@theia/core/lib/common/event';
+import { EditorManager, EditorWidget } from '@theia/editor/lib/browser';
+import { PreferenceService } from '@theia/core/lib/browser';
+import { QuickPickService } from '@theia/core/lib/common/quick-pick-service';
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
+import { TaskConfigurationModel } from './task-configuration-model';
+import { TaskCustomization, TaskConfiguration } from '../common/task-protocol';
+import { WorkspaceVariableContribution } from '@theia/workspace/lib/browser/workspace-variable-contribution';
+import { FileSystem, FileSystemError } from '@theia/filesystem/lib/common';
+import { FileChange, FileChangeType } from '@theia/filesystem/lib/common/filesystem-watcher-protocol';
+import { PreferenceConfigurations } from '@theia/core/lib/browser/preferences/preference-configurations';
+
+@injectable()
+export class TaskConfigurationManager {
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @inject(EditorManager)
+    protected readonly editorManager: EditorManager;
+
+    @inject(QuickPickService)
+    protected readonly quickPick: QuickPickService;
+
+    @inject(FileSystem)
+    protected readonly filesystem: FileSystem;
+
+    @inject(PreferenceService)
+    protected readonly preferences: PreferenceService;
+
+    @inject(PreferenceConfigurations)
+    protected readonly preferenceConfigurations: PreferenceConfigurations;
+
+    @inject(WorkspaceVariableContribution)
+    protected readonly workspaceVariables: WorkspaceVariableContribution;
+
+    protected readonly onDidChangeEmitter = new Emitter<void>();
+    readonly onDidChange: Event<void> = this.onDidChangeEmitter.event;
+
+    protected readonly onChangedTaskConfigEmitter = new Emitter<FileChange>();
+    readonly onChangedTaskConfig: Event<FileChange> = this.onChangedTaskConfigEmitter.event;
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        this.updateModels();
+        this.preferences.onPreferenceChanged(e => {
+            if (e.preferenceName === 'tasks') {
+                this.updateModels();
+            }
+        });
+    }
+
+    protected readonly models = new Map<string, TaskConfigurationModel>();
+    protected updateModels = debounce(async () => {
+        const roots = await this.workspaceService.roots;
+        const toDelete = new Set(this.models.keys());
+        for (const rootStat of roots) {
+            const key = rootStat.uri;
+            toDelete.delete(key);
+            if (!this.models.has(key)) {
+                const model = new TaskConfigurationModel(key, this.preferences);
+                model.onDidChange(() => this.onChangedTaskConfigEmitter.fire({ uri: key, type: FileChangeType.ADDED }));
```

Comment:
Should not it be `UPDATED`?

Reply:
i am not sure which (`ADDED` vs `UPDATED`) is more appropriate. to be honest i feel neither is 100% correct :)

when a new project folder is added to the workspace, from the taskManager's perspective, task configs from that new project folder are "ADDED" to the workspace. However, from the workspace's perspective, it is an `UPDATE` to the pool of task configs. 

i am ok with changing `ADDED` to `UPDATED` - it has no impact to the end effect, because subscriber of this event only cares about "delete" vs "not-a-delete".

Reply:
it also strange that we use `FileChangeEvent` here, then notion of files is lost already behind the preference service. Maybe we need another event? `TaskConfigChange` which has only an uri and then `TaskConfiguraionManager` has a method to check whether a config exist for such uri?